src/encoder.rs

```diff
@@ -1,26 +1,32 @@
-extern crate crc32fast;
-extern crate deflate;
+mod animation;
+mod image;
 
-use std::borrow::Cow;
-use std::error;
-use std::fmt;
-use std::io::{self, Read, Write};
-use std::mem;
-use std::result;
+pub use animation::{AnimationEncoder, FrameEncoder, FrameWriter};
 
-use crc32fast::Hasher as Crc32;
+use io::Write;
+use std::{error, fmt, io, result};
 
-use crate::chunk::{self, ChunkType};
-use crate::common::{BitDepth, BytesPerPixel, ColorType, Compression, Info, ScaledFloat};
+use deflate::write::ZlibEncoder;
+
+use crate::common::{BitDepth, BytesPerPixel, ColorType, Compression};
 use crate::filter::{filter, FilterType};
-use crate::traits::WriteBytesExt;
 
 pub type Result<T> = result::Result<T, EncodingError>;
 
 #[derive(Debug)]
 pub enum EncodingError {
     IoError(io::Error),
-    Format(Cow<'static, str>),
+    ZeroWidth,
+    ZeroHeight,
+    ZeroFrames,
+    Invalid(BitDepth, ColorType),
+    WrongDataSize(usize, usize),
+    AlreadyFinished,
+    MissingPalette,
+    EmptyPalette,
+    WrongPaletteLength,
+    MissingFrames(u32),
+    FrameOutOfBounds(u32, u32),
```

Comment:
The error was reworked independently from this PR. Don't throw in another kitchen sink of changes.

Reply:
Oh ok, didn't think that would need another PR, I'm going to revert those changes

Reply:
It's not so much that new errors would need another PR, it's that the style of public variants was clumsy and needed to be changed anyways. If you rebase to the new error design feel free to add new variants to the private enums `*Kind` for errors related to APNG. You'll find that the unrelated variants were also addressed, although slightly differently.